src/arrowc/schema_view.c

```diff
@@ -0,0 +1,510 @@
+  schema_view->data_type = data_type;
+  schema_view->storage_data_type = data_type;
+  schema_view->n_buffers = 2;
+  schema_view->validity_buffer_id = 0;
+  schema_view->data_buffer_id = 1;
+}
+
+ArrowErrorCode ArrowSchemaViewSetStorageType(struct ArrowSchemaView* schema_view,
+                                             const char* format, const char** format_end,
+                                             struct ArrowError* error) {
+  schema_view->validity_buffer_id = -1;
+  schema_view->offset_buffer_id = -1;
+  schema_view->large_offset_buffer_id = -1;
+  schema_view->data_buffer_id = -1;
+  schema_view->type_id_buffer_id = -1;
+  *format_end = format;
+
+  // needed for decimal parsing
+  const char* parse_start;
+  char* parse_end;
+
+  switch (format[0]) {
+    case 'n':
+      schema_view->storage_data_type = ARROWC_TYPE_NA;
+      schema_view->n_buffers = 0;
+      *format_end = format + 1;
+      return ARROWC_OK;
+    case 'b':
+      ArrowSchemaViewSetPrimitive(schema_view, ARROWC_TYPE_BOOL);
+      *format_end = format + 1;
+      return ARROWC_OK;
+    case 'c':
+      ArrowSchemaViewSetPrimitive(schema_view, ARROWC_TYPE_INT8);
+      *format_end = format + 1;
+      return ARROWC_OK;
+    case 'C':
+      ArrowSchemaViewSetPrimitive(schema_view, ARROWC_TYPE_UINT8);
+      *format_end = format + 1;
+      return ARROWC_OK;
+    case 's':
+      ArrowSchemaViewSetPrimitive(schema_view, ARROWC_TYPE_INT16);
+      *format_end = format + 1;
+      return ARROWC_OK;
+    case 'S':
+      ArrowSchemaViewSetPrimitive(schema_view, ARROWC_TYPE_UINT16);
+      *format_end = format + 1;
+      return ARROWC_OK;
+    case 'i':
+      ArrowSchemaViewSetPrimitive(schema_view, ARROWC_TYPE_INT32);
+      *format_end = format + 1;
+      return ARROWC_OK;
+    case 'I':
+      ArrowSchemaViewSetPrimitive(schema_view, ARROWC_TYPE_UINT32);
+      *format_end = format + 1;
+      return ARROWC_OK;
+    case 'l':
+      ArrowSchemaViewSetPrimitive(schema_view, ARROWC_TYPE_INT64);
+      *format_end = format + 1;
+      return ARROWC_OK;
+    case 'L':
+      ArrowSchemaViewSetPrimitive(schema_view, ARROWC_TYPE_UINT64);
+      *format_end = format + 1;
+      return ARROWC_OK;
+    case 'e':
+      ArrowSchemaViewSetPrimitive(schema_view, ARROWC_TYPE_HALF_FLOAT);
+      *format_end = format + 1;
+      return ARROWC_OK;
+    case 'f':
+      ArrowSchemaViewSetPrimitive(schema_view, ARROWC_TYPE_FLOAT);
+      *format_end = format + 1;
+      return ARROWC_OK;
+    case 'g':
+      ArrowSchemaViewSetPrimitive(schema_view, ARROWC_TYPE_DOUBLE);
+      *format_end = format + 1;
+      return ARROWC_OK;
+
+    // decimal
+    case 'd':
+      if (format[1] != ':' || format[2] == '\0') {
+        ArrowErrorSet(error, "Expected ':precision,scale[,bitwidth]' following 'd'",
+                      format + 3);
+        return EINVAL;
+      }
+
+      parse_start = format + 2;
+      schema_view->decimal_precision = strtol(parse_start, &parse_end, 10);
```

Comment:
As a TODO we should switch to locale-agnostic functions for parsing and generating format strings.